### PR TITLE
dynamicconfig: Add dynamic config cell module

### PR DESCRIPTION
**Pull request by ovidiutirla**
The agent will monitor the designated ConfigMap and automatically synchronize any changes with its in-memory database, StateDB. Create a new cell in the Agent that watches the configured ConfigMap and creates a StateDB table to store the key-value pairs. The Agent cell will expose the read-only version of the StateDB table.

Please ignore the first commit as it's a squash from https://github.com/cilium/cilium/pull/34386

Related: https://github.com/cilium/cilium/issues/34101

```release-note
Add dynamic configuration for agents through ConfigMap.
```

### CODE REVIEW

File: pkg/dynamicconfig/cell.go
Comment (joestringer) [inline pkg/dynamicconfig/cell.go:54]: SGTM. I do wonder about how we make the user experience around this clear. I'm thinking for example, maybe a user finds this configuration option, enables it. But then in practice none of the options can be dynamically modified until we start to incrementally introduce support for those things.

I know that on the CFP we had some discussion on this, and there's a tradeoff - I was arguing that there should still be a single configmap so that users don't have to first decide where to configure the options and then figure out the right configmap/resource to configure. On the other hand, when restart is required to modify a setting, it would be nice if the user doesn't get caught out by this, but rather Cilium "just works" to reconfigure the options.

Perhaps a solution to this is to integrate better with the Helm option `rollOutCiliumPods`, so that when the user specifies this option, that enables dynamic config update as well. If Cilium determines that restart is required, it would restart the agents. If restart is not required, ideally the same option would not restart the agents. So then the user just configures one high level option to enable dynamic config, and the details about how the dynamic config gets applied is up to Cilium. Cilium just figures out the most efficient way to apply the config based on the implementation. If this proposal isn't compatible with the implementation details of the rolloutpods setting then it could alternatively be a new Helm flag for dynamic config, we could explore that.

Related, we might consider setting this flag to hidden to reduce the likelihood that a user discovers this flag and then misunderstands the setting, sets it, then is surprised that by itself it doesn't make all the settings dynamic. Another mitigation might be something like the following (would also require a docs update for dynamic configs to close the loop):

```suggestion
	flags.Bool("enable-dynamic-config", c.EnableDynamicConfig, "Enables support for dynamic agent config for supported configurations (see https://docs.cilium.io/en/stable/configuration/dynamic-config/ for a full list)")
```

Comment (ovidiutirla) [reply pkg/dynamicconfig/cell.go:54]: Indeed, totally agree with you. So I think it make sense to hide this for now. I was planning to push these changes so we can actually have some foundation that we can build on and let other devs know that they can use this in their features. 

I was planning to find a way to document this for users, I can add the page for the supported configuration on my task list and document such a page. 

I think we can extend this to support a feature where we let cilium decide if it needs restart or not, along with the DynamicLifecycle. So I think then we'll have like 3 options 1. dynamic apply the config changes, 2. FeatureManager could restart(or just start/stop) only specific cells, 3. restart the whole cilium agent.

File: pkg/dynamicconfig/table.go
Comment (joamaki) [inline pkg/dynamicconfig/table.go]: WDYT about waiting for https://github.com/cilium/cilium/pull/34386 to merge and then doing `TransformMany` and have a flat representation? This sort of nesting is bit of an anti-pattern as now you cannot watch individual keys and it's inconvenient to query.

Maybe something along these lines for the types:
```
// Key for an option within specific config map
type Key struct {
  MapName string
  Key string
}

func (k Key) String() string {
  return k.MapName + "/" + k.Key
}

type Option struct {
  Key Key
  Value string
}

var keyIndex = statedb.Index[Option, Key]{
  ...
  FromObject: func(opt Option) index.KeySet {
    return index.NewKeySet(opt.Key.String())
  }, ...
}
var ByKey = keyIndex.Query

// example usage:
var tbl statedb.Table[dynamicconfig.Option]
opt, _, watch, found := tbl.GetWatch(txn, dynamicconfig.ByKey(dynamicconfig.Key{MapName: "foo", Key: "bar"})

// Now we can watch individual keys for changes:
<-watch
```

File: pkg/dynamicconfig/table.go
Comment (joamaki) [inline pkg/dynamicconfig/table.go]: Please add a command to `cilium-dbg/cmd/statedb.go`. You'll just need a one-liner there plus `TableRow` and `TableHeader` implementation here for `DynamicConfig`.

**Review by joamaki — CHANGES_REQUESTED** [review]

File: pkg/dynamicconfig/table.go
Comment (joamaki) [inline pkg/dynamicconfig/table.go]: Are these really unique?

File: pkg/dynamicconfig/table.go
Comment (joamaki) [inline pkg/dynamicconfig/table.go]: I've been naming these `BySomething`, e.g. `ByKey`

File: pkg/dynamicconfig/table.go
Comment (christarazi) [inline pkg/dynamicconfig/table.go]: Have you also considered CiliumNodeConfig?

**Review by christarazi — COMMENTED** [review]
This is a great start, thanks for the PR!

I am wondering about two things:

1. Whether we should add support for CiliumNodeConfig while we are at it. 
2. Similar to https://github.com/cilium/cilium/pull/31101, having a status mechanism to inform the user that Cilium has detected a change in config can be very useful, especially in the meantime while we slowely add support over time for what was discussed in https://github.com/cilium/cilium/pull/34383/files#r1717310296.

For (2), the daemon can simply read from the new table that this PR adds, rather than query K8s directly. If this sounds reasonable, I think this PR would supersede https://github.com/cilium/cilium/pull/31101 and we can stop investing time into that solution.

File: pkg/dynamicconfig/table.go
Comment (ovidiutirla) [reply pkg/dynamicconfig/table.go]: Yes, the idea is to support multiple ConfigMaps including the LocalNodeConfig, I didn't have time today to rework this implementation but the goal is to refactor using the TransformMany and flatten the Map and the key would be something like (Key, Source) -> Value, and some logic to define the priority of each source, where Key is the key from the configmaps, the source is the configmap name, and Value is the key's values. 

So when we query this, we'll return the value from the ConfigMap with the highest priority.


I would prefer to rework this and have the logic for adding other ConfigMaps in this PR but adding the CiliumLocalNodeConfig to be in a later PR, I can still take care of it but I think having this piece in for now could be useful. Also having that PR as an example of how to add other ConfigMaps could facilitate later additions.

Comment (christarazi) [reply pkg/dynamicconfig/table.go]: What is your take on (2)?

Comment (ovidiutirla) [reply pkg/dynamicconfig/table.go]: That sounds like a feasible approach, I advocate for adding such a way to notify in case of config drift. I think we can still have the logic for computing the deltas in the daemon using this table. Or do you think it would fit better in the dynamicconfig pkg?

Comment (christarazi) [reply pkg/dynamicconfig/table.go]: My thinking is that we don't necessarily need to compute the diff outside of the dynamicconfig pkg since it's already implementing logic to fetch the desired state. Might as well do it in the dynamicconfig pkg. It would then just be a matter of diffing with the realized state and setting some flag (or something similar) that then the daemon can fetch in its status report to indicate that config has changed and some action is needed.

Comment (ovidiutirla) [reply pkg/dynamicconfig/table.go]: Let's continue the discussion on Slack, and I'll follow up with another PR for this part.

**Comment by ovidiutirla** [comment]
Please ignore the first commit as it's a squash from https://github.com/cilium/cilium/pull/34386.

<details>
  <summary>Testing</summary>
 
  ### One source
image

  ### One additional source

image
image



</details>

**Review by marseel — APPROVED** [review]
vendor changes lgtm

**Review by bimmlerd — APPROVED** [review]
CODEOWNERS lgtm.

**Comment by ovidiutirla** [comment]
rebased on main

File: pkg/dynamicconfig/cell.go
Comment (chancez) [reply pkg/dynamicconfig/cell.go:54]: > Perhaps a solution to this is to integrate better with the Helm option rollOutCiliumPods,

@joestringer I suspect it will be difficult to do this, as the way `rollOutCiliumPods` works is by hashing the entire rendered configmap and creating checksum annotation on the pod that will change when the configMap changes so the pod template changes when the configmap does, causing the daemonset to roll out the pods.

To make it integrate in the way your suggesting, will require manually constructing the input values for the checksum calculation, based on whether or not that value is dynamically reloadable. Not to mention, I suspect this behavior would be somewhat confusing. From a devops perspective, I might prefer to have it restart the pods, even if the changed options are dynamically reloadedable when I use `rollOutCiliumPods`. I think they're similar, but distinct features.

File: pkg/dynamicconfig/table.go
Comment (tommyp1ckles) [inline pkg/dynamicconfig/table.go]: I think [state.Collect(...)](https://pkg.go.dev/github.com/cilium/statedb#Collect) already does this.

File: install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml
Comment (tommyp1ckles) [inline install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml]: By the principle of least privilege, I wonder if we should only enable this if the dynamicconfig is enabled?

File: pkg/dynamicconfig/cell.go
Comment (tommyp1ckles) [reply pkg/dynamicconfig/cell.go:54]: Just an idea, perhaps in the future we could set a health status to degraded, or emit a specific warning on a Agent pod that detects it needs a restart to have up to date config?  If enabled, would provide a way for users to be notified about whether they have out of sync Agents.

**edit**: Think Chris already proposed this below.

File: pkg/dynamicconfig/table.go
Comment (tommyp1ckles) [inline pkg/dynamicconfig/table.go]: Question:  not sure if this would actually be meaningful, could we ever expect to see default (i.e. 0) priorities, in such a case the ordering would be undefined.  Would it make sense to brake ties using the key alphabetically?

File: install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml
Comment (joestringer) [reply install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml]: Something else interesting to me on this front - Cilium already receives configmap updates for its own configmap somehow.. do we need to grant access to all configmaps or is the current mechanism sufficient?

File: pkg/dynamicconfig/table.go
Comment (ovidiutirla) [reply pkg/dynamicconfig/table.go]: There is a plan to deprecate this anyway, part of: https://github.com/cilium/statedb/pull/40. 

Replaced with `statedb.Collect()`, Thanks!

File: pkg/dynamicconfig/table.go
Comment (ovidiutirla) [reply pkg/dynamicconfig/table.go]: There shouldn't be a case where the priorities are equal considering the source was added properly. Sorting alphabetically would just add more confusion in the code IMO.  Each time a source is added, its corresponding priority needs to be added [here](https://github.com/cilium/cilium/pull/34383/files#diff-e93664a938a87894058f8d6c041422e609d6d271c8828b127a12226bcf07e00dR22)

File: install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml
Comment (ovidiutirla) [reply install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml]: Without this clusterrole I get the following and the pod is not able to get/watch the configmap:
```
time="2024-08-19T21:31:55Z" level=warning msg="github.com/cilium/cilium/pkg/k8s/statedb.go:336: failed to list <unspecified>: configmaps \"cilium-config\" is forbidden: User \"system:serviceaccount:kube-system:cilium\" cannot list resource \"configmaps\" in API group \"\" at the cluster scope" subsys=klog
```

Indeed, you are right @joestringer, I've updated the watcher to only look in `kube-system` namespace and looks like it have access there without the clusterrole.

I'll revert the clusterrole changes. Thanks for spotting @tommyp1ckles ! Nice one!

File: pkg/dynamicconfig/table.go
Comment (tommyp1ckles) [reply pkg/dynamicconfig/table.go]: sounds good thanks for the clarification

File: pkg/dynamicconfig/table_test.go
Comment (tommyp1ckles) [inline pkg/dynamicconfig/table_test.go:50]: nit: I'd recommend using the assert.Eventually family of functions for this, it leads to easier to read code and tends to be more robust (although in this case this seems fine).

File: pkg/dynamicconfig/table_test.go
Comment (tommyp1ckles) [inline pkg/dynamicconfig/table_test.go]: spelling: stoping -> stopping

**Review by tommyp1ckles — APPROVED** [review]
couple minor things but @cilium/sig-k8s changes look good.

File: pkg/dynamicconfig/table_test.go
Comment (ovidiutirla) [reply pkg/dynamicconfig/table_test.go:50]: Indeed, I agree, but I wanted to include the test for the watcher (chan) to make sure the index gets invalidated by the different sources.

File: pkg/dynamicconfig/cell.go
Comment (ovidiutirla) [reply pkg/dynamicconfig/cell.go:54]: Yes, it's on my radar to cover this part mentioned by Chris, I'll make sure to add both as reviewers.

**Comment by ovidiutirla** [comment]
rebase on main

**Comment by squeed** [comment]
I'm a bit concerned that this duplicates much of work from [CiliumNodeConfig](https://docs.cilium.io/en/stable/configuration/per-node-config/#ciliumnodeconfig-objects). We already have code that resolves configuration -- albeit on startup -- and supplies key-value pairs. It even has a node selection mechanism, so configuration can be selectively applied based on labels.

I think this is a good idea, but it should use the already-existing logic that resolves configuration from ConfigMaps, the CNC, and Node annotations.

**Comment by ovidiutirla** [comment]
> I'm a bit concerned that this duplicates much of work from [CiliumNodeConfig](https://docs.cilium.io/en/stable/configuration/per-node-config/#ciliumnodeconfig-objects). We already have code that resolves configuration -- albeit on startup -- and supplies key-value pairs. It even has a node selection mechanism, so configuration can be selectively applied based on labels.
> 
> I think this is a good idea, but it should use the already-existing logic that resolves configuration from ConfigMaps, the CNC, and Node annotations.

Thanks for the feedback @squeed. 

You mean to take the config-map name and namespace from the `--config-sources='config-map:kube-system/cilium-config'"`?

This PR currently only adds support for the base cilium-config ConfigMap, I plan to have a subsequent PR to address the `CiliumNodeConfig` concern as another source of configuration.

File: pkg/dynamicconfig/table_test.go
Comment (joestringer) [reply pkg/dynamicconfig/table_test.go:50]: @ovidiutirla I don't quite understand your response here - We generally avoid timer-based waits in Cilium tests, because they're racy and can make the tests unreliable in environments with high CPU usage, which is not uncommon in CI. Is there a way to implement this test without referencing discrete time units like `time.Second`? `assert.Eventually` is a common pattern, or otherwise using channels.

Comment (ovidiutirla) [reply pkg/dynamicconfig/table_test.go:50]: The time is used as a timeout, the channel is closed in the happy path when the values is updated in the table.

Comment (joestringer) [reply pkg/dynamicconfig/table_test.go:50]: OK. I guess if it turns out this is unreliable then we can always revisit it.

**Comment by ovidiutirla** [comment]
Agreed to have a follow-up PR for handing the CNCs and another one for the drift checker. The goal is to store in the stateDB table all the configs that we currently use so the drift checker should report no diffs. Ideally, the goal is to try to re-use the existing `pkg/resolver` logic and have only one implementation for sources.

File: pkg/dynamicconfig/cell.go
Comment (joamaki) [inline pkg/dynamicconfig/cell.go:29]: Add a short comment for `var Cell` that describes what it does and what it provides. Essentially I'd mention that it reflects config maps into the "Table[DynamicConfig]" and that this table can be used to access them. E.g. the reader should at a glance have a good idea what's going on here without looking further. Having a code example or a pointer to an example for how you use what this provides is also helpful.

Also the module title doesn't need to repeat the ID. The idea is to use the id and title to visualize things and it'll appear as "<id>: <title>", so the title should be more of a description like "Reflects ConfigMaps to the DynamicConfig table", so then when we show the module it reads: "cilium-agent-dynamic-config: Reflects ConfigMaps to the DynamicConfig table"

File: pkg/dynamicconfig/table.go
Comment (joamaki) [inline pkg/dynamicconfig/table.go]: `Name` here and below out of date. I'd just have `key` here, and maybe `name` below (these aren't actually yet used for anything, but eventually I think we'll have some command or documentation for all the tables and their indexes...).

File: pkg/dynamicconfig/table.go
Comment (joamaki) [inline pkg/dynamicconfig/table.go]: This should be just `ByKey` and below it should be `ByName`

File: pkg/dynamicconfig/table.go
Comment (joamaki) [inline pkg/dynamicconfig/table.go]: Hm slightly inconsistent that `GetKey` returns `(DynamicConfig, bool)` and here we don't have the `bool`. Begs the question whether `GetKey` needs it either. I would make these consistent to each other.

File: pkg/dynamicconfig/table.go
Comment (joamaki) [inline pkg/dynamicconfig/table.go]: Hmm, for me intuitively a low number means high priority. That's also how we've dealt with priority in other places (for example `pkg/bgpv1/manager/reconciler/reconciler.go`). So I find this surprising and unconventional. Not sure if it's just me though :D

File: pkg/dynamicconfig/table_test.go
Comment (joamaki) [inline pkg/dynamicconfig/table_test.go]: Do we need this `waitFor` construction? This should be doable all from a single goroutine, e.g. you do `upsertDummyEntry` and then you block on the watch channel and validate. It's also not advisable to use `testing.T` from other goroutines.

**Review by joamaki — CHANGES_REQUESTED** [review]

File: pkg/dynamicconfig/table_test.go
Comment (ovidiutirla) [reply pkg/dynamicconfig/table_test.go]: Yes you are right. removed the waitFor. Though I've kept the select in case the test is failing to timeout after 2 seconds? Do you see a better way to do it?

File: pkg/dynamicconfig/cell.go
Comment (ovidiutirla) [reply pkg/dynamicconfig/cell.go:29]: I've added a short description

**Review by joamaki — APPROVED** [review]

**Comment by squeed** [comment]
@ovidiutirla please do file a release-blocking issue to add CNC support. I don't want to have any chance of shipping v1.17 with configuration divergence.

**Comment by ovidiutirla** [comment]
I've created https://github.com/cilium/cilium/issues/34541. Can you please add the relevant labels and assign it to me?(I don't have access)

**Comment by ldelossa** [comment]
Just a heads up, this PR is marked `ready-to-merge` but is blocked until this: https://github.com/cilium/cilium/pull/34383#discussion_r1717225633 conversation is resolved. Can we either resolve it or remove the `ready-to-merge` label?

**Comment by ovidiutirla** [comment]
> Just a heads up, this PR is marked `ready-to-merge` but is blocked until this: [#34383 (comment)](https://github.com/cilium/cilium/pull/34383#discussion_r1717225633) conversation is resolved. Can we either resolve it or remove the `ready-to-merge` label?

I wasn't aware that unresolved comments are blocking, I resolved it and will create follow-up issues to track the discussion. thanks.

**Comment by ldelossa** [comment]
No problem @ovidiutirla :)

**Pull request merged**
